### ISSUE TITLE
Missing XMLHttpRequest on node

### ISSUE THREAD

**Issue by gregkare** [remotestorage/remotestorage.js#1034]
Extracted from #1032 

```
$ node
> > var RemoteStorage = require('remotestoragejs');
... ReferenceError: XMLHttpRequest is not defined
...     at Object.<anonymous> (./node_modules/remotestoragejs/release/remotestorage.js:28:4332)
...     at t (./node_modules/remotestoragejs/release/remotestorage.js:2:352)
...     at Object.e.exports (./node_modules/remotestoragejs/release/remotestorage.js:28:53)
...     at t (./node_modules/remotestoragejs/release/remotestorage.js:2:352)
...     at Object.e.defaultValue (./node_modules/remotestoragejs/release/remotestorage.js:3:16231)
...     at t (./node_modules/remotestoragejs/release/remotestorage.js:2:352)
...     at Object.n (.node_modules/remotestoragejs/release/remotestorage.js:4:8757)
...     at t (./node_modules/remotestoragejs/release/remotestorage.js:2:352)
...     at Object.<anonymous> (./node_modules/remotestoragejs/release/remotestorage.js:2:472)
...     at t (./node_modules/remotestoragejs/release/remotestorage.js:2:352)
```

Including a dependency on a XMLHttpRequest library (such as xmlhttprequest) and instantiating a `XMLHttpRequest` variable before requiring `remotestorage` fixes it.

We used to require xhr2 as the XMLHttpRequest variable (c5d78f7d) and still have a dependency on it in package.json, but we're not using it in the code at the moment.

How should we fix this?

**Comment by skddc** [comment]
I was trying to find out how to best handle this for about an hour, and I'm none the wiser now. The options I currently see:

### 1. Point the package's `main` back to an adapted `node-main.js` and use the not-entirely-but-almost-standard `browser` property for the browser build alone.

#### Pros:
* Node.js works out of the box by just requiring the module
* Only one package to know about and maintain

#### Cons:
* Browser builds might not work out of the box, but some (or most?) of them will (e.g. Browserify and Webpack would work afaics)

### 2. Create a second npm package, which is nothing but the current `node-main.js` and depends on `remotestoragejs`, could be `remotestoragejs-node`. Leave the UMD build as default in the main package and remove everything node-specific.

#### Pros:
* Node.js works out of the box with the node-specific package
* Browser packaging works out of the box with the normal/main package

#### Cons:
* Need to know about second package as user
* Need to maintain a second package (keep versions in sync and such)

### 3. Remove the `node-main.js`, but also don't provide an automatic solution, and instead add documentation for adding `xhr2` and the `readBinaryData` fix to one's own code.

#### Pros:
* Browser packaging works out of the box with the normal/main package
* No extra code or hacks for node in main package

#### Cons:
* Always need to know about and add custom for usage with node.js

### 4. Some way of environment detection in the UMD build, that I haven't presented and am not sure is viable (but could be)

#### Pros:
* Node.js works out of the box with the node-specific package
* Browser packaging works out of the box in all cases

#### Cons:
* Not sure if possible/viable (but could be)
* Increased build size if everyone has to have `xhr2` code in the build (which might not have to be the case)

---

I might have missed something, but these are the options I can think of right now.

**Comment by galfert** [comment]
Thanks for your research and the summary. I don't have much insight about this either and haven't formed an opinion yet.

Maybe @lesion or @silverbucket can chime in?

**Comment by skddc** [comment]
... btw, the `readBinaryData` fix in `node-main.js` can't work like this anymore, because of the new library layout.

**Comment by skddc** [comment]
Made some progress. See linked PR.

**Comment by skddc** [comment]
Update: @galfert is working on the `readBinaryData` fix.

**Comment by galfert** [comment]
LOL, I just found out that the `readBinaryData` function isn't even used by Wireclient anymore since another fix for binary data (615dcd0598cce9b).

So we can just remove any reference to it from the node index.js as well.

It is still used by Dropbox though. Looks like we just forgot about that back then. I'll check if we can also just remove it there as well and get rid of that function completely.

**Comment by skddc** [comment]
![lol-picard](https://user-images.githubusercontent.com/842/32059656-10006e8e-ba6d-11e7-9ef6-2ca3243f0921.gif)
